### DISCUSSION TITLE
Feature request: Reusable interactive components

### DISCUSSION THREAD

**Discussion by dialmove** [uxbox/uxbox#709] [Ideas]
Current approach to interaction is by switching full screens - one screen or "frame" for each Artboard. Following a link swaps the whole screen.

Yet if a group of Artboards (connected with interactions between them) could be shared as a component, this would allow to prototype finer interactions. Inserting such component as a nested object in a larger frame creates an interactive widget.

With this feature, following a link inside the component would swap just the widget status, changing it to a different Artboard in the component and placing it at the same location, leaving the rest of the page unchanged.

This allows simulation of widgets like drop-downs, tab groups, date pickers... The same interaction could be built once and reused in several pages.

**Comment by dialmove** [comment]
I'm reviewing other feature requests in this discussion, and I've found [this one for component variants](https://github.com/penpot/penpot/discussions/534).

I believe that this "component variants" feature could be _prototyped_ with the reusable interactive components I propose:
- Create 12 Artboards for the 12 states of the button
 ![image](https://user-images.githubusercontent.com/11275573/106670572-83357b80-65ad-11eb-8286-d1fea0c31466.png)

- Create the control panels inside the 12 artboards, with labels and switchers for the button attributes
> ![image](https://user-images.githubusercontent.com/11275573/106671711-3488e100-65af-11eb-9e6b-f8d3c81180c1.png)

- Link the switchers with "Navigate to" interactions leading to the button state they enable
- Wrap the whole artwork group as a component. The component would need a "viewport" focusing only on the button, hiding the control panels.

This button component could be instanced in other designs as a button with several states. Changing the "default" state of the instanced component would change its state.

**Comment by dialmove** [comment]
This feature is inspired by the ["States" feature in Infragistics Indigo Studio](http://help.indigodesigned.com/about-indigo-studio/advice-on-prototyping-with-indigo), which has a [powerful model for prototyping interactions](http://help.indigodesigned.com/working-with-interactions/understanding-screen-interactions-and-states).

Infragistics tool has a wizard for defining states of interactive components (this is not necessary in Penpot for the feature to work if components are created using standard Artworks):
> ![image](http://help.indigodesigned.com/images/Understanding_Screen_Interactions_and_States_3.png)

State changes are tied to an event triggered in the View mode, which changes the component to a different defined state. In Penpot, this could be equivalent to navigating between artworks in the component:
> ![image](http://help.indigodesigned.com/images/Understanding_Screen_Interactions_and_States_4.png)

**Comment by jcklpe** [comment]
I have previously written a request for this feature under "microinteractions" which I think is worth splitting up from "variants"

- variants: https://github.com/penpot/penpot/issues/768
- microinteractions: https://github.com/penpot/penpot/issues/769

There's also a forum post that was posted this summer on the subject: 
https://github.com/penpot/penpot/discussions/709

Just posting these links here to try and maximize the network interlinking of these discussions. Penpot should probably think through a little more clearly where things like feature requests, discussions, etc take place, since right now there's at least 3 places they could reside (forums, github discussions, github issues)  Possibly 4 if we count taiga. 

Oh yes there's also a taiga thing thing for this on the roadmap:
- https://tree.taiga.io/project/penpot/us/527

Just noticed that the taiga story is closed and archived. Perhaps @elhombretecla can clarify.
